jormungandr/src/explorer/mod.rs

```diff
@@ -66,112 +89,303 @@ impl Explorer {
+    future::poll_fn(move || Ok(lock.poll_lock()))
+}
+
+fn apply_block_to_transactions(transactions: Transactions, block: &Block) -> Result<Transactions> {
+    let block_id = block.id();
+    let ids = block
+        .contents
+        .iter()
+        .filter(|fragment| is_transaction(fragment))
+        .map(|fragment| fragment.id());
+
+    let mut transactions = transactions;
+    for id in ids {
+        transactions = transactions
+            .insert(id, block_id)
+            .map_err(|_| ErrorKind::TransactionAlreadyExists(format!("{}", id)))?;
+    }
+
+    Ok(transactions)
+}
+
+fn apply_block_to_blocks(blocks: Blocks, block: &Block) -> Result<Blocks> {
+    let block_id = block.id();
+    blocks
+        .insert(block_id, (*block).clone())
+        .map_err(|_| Error::from(ErrorKind::BlockAlreadyExists(format!("{}", block_id))))
+}
+
+fn apply_block_to_addresses(addresses: Addresses, block: &Block) -> Result<Addresses> {
+    // FIXME: Probably there is a cleaner way to do this
+    let txs = block.contents.iter().filter_map(|fragment| match fragment {
+        Fragment::Transaction(auth_tx) => Some((
+            fragment.id(),
+            &auth_tx.transaction.inputs,
+            &auth_tx.transaction.outputs,
+        )),
+        Fragment::OwnerStakeDelegation(auth_tx) => Some((
+            fragment.id(),
+            &auth_tx.transaction.inputs,
+            &auth_tx.transaction.outputs,
+        )),
+        Fragment::StakeDelegation(auth_tx) => Some((
+            fragment.id(),
+            &auth_tx.transaction.inputs,
+            &auth_tx.transaction.outputs,
+        )),
+        Fragment::PoolRegistration(auth_tx) => Some((
+            fragment.id(),
+            &auth_tx.transaction.inputs,
+            &auth_tx.transaction.outputs,
+        )),
+        Fragment::PoolManagement(auth_tx) => Some((
+            fragment.id(),
+            &auth_tx.transaction.inputs,
+            &auth_tx.transaction.outputs,
+        )),
+        _ => None,
+    });
+
+    let mut addresses = addresses;
+
+    for (txid, inputs, outputs) in txs {
+        for output in outputs {
+            addresses = addresses
+                .insert_or_update(
+                    output.address.clone(),
+                    [txid].iter().cloned().collect(),
+                    |set| {
+                        let mut new_set: HashSet<FragmentId> = set.iter().cloned().collect();
+                        new_set.insert(txid);
+                        std::result::Result::<_, Infallible>::Ok(Some(new_set))
+                    },
+                )
+                // FIXME: I still don't know when a insert or update can fail
+                .unwrap();
+        }
+        for input in inputs {
+            match input.to_enum() {
+                InputEnum::AccountInput(id, _value) => {
+                    // TODO: How do I get an Address from an AccountIdentifier?
+                    // Can I do it without knowing the Discrimination?
```

Comment:
easy: from the block0/genesis block.

But there is another problem/question for you is that you also don't know what `prefix` to use for the address in its bech32 representation.

a question for @vincenthz to help with maybe.

From my point of view we can either:

* add this in the configuration file of the node (under explorer: address prefix);
* let the user decide from the GraphQL query;
* just ignore it and return the public keys only?